rainier-tests/src/test/scala/com/stripe/rainier/compute/RealTest.scala

```diff
@@ -4,29 +4,59 @@ import org.scalatest._
 import com.stripe.rainier.core._
 import scala.util.{Try, Success, Failure}
 class RealTest extends FunSuite {
-  def run(description: String)(fn: Real => Real): Unit = {
+  def run(description: String, testDeriv: Boolean = true)(
+      fn: Real => Real): Unit = {
     test(description) {
       val x = new Variable
       val result = fn(x)
+      val deriv = result.gradient.head
+
+      def evalAt(d: Double): Double = Try { fn(Constant(d)) } match {
+        case Success(Infinity)                 => 1.0 / 0.0
+        case Success(NegInfinity)              => -1.0 / 0.0
+        case Success(Constant(bd))             => bd.toDouble
+        case Failure(_: ArithmeticException)   => 0.0 / 0.0
+        case Failure(_: NumberFormatException) => 0.0 / 0.0
+        case x                                 => sys.error("Non-constant value " + x)
+      }
+
       val c = Compiler(200, 100).compile(List(x), result)
+      val dc = Compiler(200, 100).compile(List(x), deriv)
       List(1.0, 0.0, -1.0, 2.0, -2.0, 0.5, -0.5).foreach { n =>
-        val constant = Try { fn(Constant(n)) } match {
-          case Success(Infinity)                 => 1.0 / 0.0
-          case Success(NegInfinity)              => -1.0 / 0.0
-          case Success(Constant(bd))             => bd.toDouble
-          case Failure(_: ArithmeticException)   => 0.0 / 0.0
-          case Failure(_: NumberFormatException) => 0.0 / 0.0
-          case x                                 => sys.error("Non-constant value " + x)
-        }
+        val constant = evalAt(n)
         val eval = new Evaluator(Map(x -> n))
         val withVar = eval.toDouble(result)
         assertWithinEpsilon(constant, withVar, s"[c/ev, n=$n]")
         val compiled = c(Array(n))
         assertWithinEpsilon(withVar, compiled, s"[ev/ir, n=$n]")
+
+        // derivatives of automated differentiation vs numeric differentiation
+        if (testDeriv && n != 0.0) { // exclude 0 since a lot of the test inputs are not defined or not continuous there
+          val dx = 10E-6
+          val numDiff = (evalAt(n + dx) - evalAt(n - dx)) / (dx * 2)
+          val diffWithVar = eval.toDouble(deriv)
+          assertAlmostEqual(numDiff,
+                            diffWithVar,
+                            s"[numDiff/diffWithVar, n=$n]")
+          val diffCompiled = dc(Array(n))
+          assertAlmostEqual(diffWithVar,
+                            diffCompiled,
+                            s"[diffWithVar/diffCompiled, n=$n]")
+        }
       }
     }
   }
 
+  def assertAlmostEqual(x: Double,
```

Comment:
Is there a reason to keep both this and `assertWithinEpsilon`?

Reply:
Originally, assertWithinEpsilon was not tolerant enough for the inaccuracy of numeric differentiation. Two sided differentiation now provides enough accuracy and I have changed the test to use assertWithinEpsilon everywhere.